server/src/infra/tcp_server.rs

```diff
@@ -12,22 +12,31 @@ use crate::{
         tcp_backend_handler::*,
     },
 };
-use actix_files::{Files, NamedFile};
+use actix_files::Files;
 use actix_http::{header, HttpServiceBuilder};
 use actix_server::ServerBuilder;
 use actix_service::map_config;
 use actix_web::{dev::AppConfig, guard, web, App, HttpResponse, Responder};
 use anyhow::{Context, Result};
 use hmac::Hmac;
+use itertools::Itertools;
 use sha2::Sha512;
 use std::collections::HashSet;
-use std::path::PathBuf;
 use std::sync::RwLock;
 use tracing::info;
 
-async fn index() -> actix_web::Result<NamedFile> {
-    let path = PathBuf::from(r"app/index.html");
-    Ok(NamedFile::open(path)?)
+async fn index<Backend>(data: web::Data<AppState<Backend>>) -> actix_web::Result<impl Responder> {
+    let mut file = std::fs::read_to_string(r"./app/index.html")?;
+
+    if data.server_url.path() != "/" {
+        file = file
+            .split_terminator("<base href=\"/\">")
```

Comment:
I think you can use https://doc.rust-lang.org/std/string/struct.String.html#method.replace instead.